public/app/features/variables/adhoc/actions.ts

```diff
@@ -34,82 +36,104 @@ export const applyFilterFromTable = (options: AdHocTableOptions): ThunkResult<vo
     if (!variable) {
       dispatch(createAdHocVariable(options));
       variable = getVariableByOptions(options, getState());
+      if (!variable) {
+        return;
+      }
     }
 
     const index = variable.filters.findIndex((f) => f.key === options.key && f.value === options.value);
 
     if (index === -1) {
       const { value, key, operator } = options;
       const filter = { value, key, operator, condition: '' };
-      return await dispatch(addFilter(variable.id, filter));
+      return await dispatch(addFilter(toKeyedVariableIdentifier(variable), filter));
     }
 
     const filter = { ...variable.filters[index], operator: options.operator };
-    return await dispatch(changeFilter(variable.id, { index, filter }));
+    return await dispatch(changeFilter(toKeyedVariableIdentifier(variable), { index, filter }));
   };
 };
 
-export const changeFilter = (id: string, update: AdHocVariabelFilterUpdate): ThunkResult<void> => {
+export const changeFilter = (
+  identifier: KeyedVariableIdentifier,
+  update: AdHocVariabelFilterUpdate
+): ThunkResult<void> => {
   return async (dispatch, getState) => {
-    const variable = getVariable(id, getState());
-    dispatch(filterUpdated(toVariablePayload(variable, update)));
-    await dispatch(variableUpdated(toVariableIdentifier(variable), true));
+    const variable = getVariable(identifier, getState());
+    dispatch(toKeyedAction(identifier.stateKey, filterUpdated(toVariablePayload(variable, update))));
+    await dispatch(variableUpdated(toKeyedVariableIdentifier(variable), true));
```

Comment:
Yes, you're absolutely correct. We pass identifiers around to be 100% sure that we don't pass any stale objects. I can look into how this could be further simplified after this PR. Thanks for bringing it up 🙏

Reply:
Tracking this in another issue: #44766